File: Data Structure/Array/Array Data Structure Guide/Basic Problems on Array/Alternate elements of an array/Recursive Approach/method1.java
// Recursive Java Program to print alternate elements
// of the array

import java.util.ArrayList;

class Main {
    
    // Recursive function to store all alternate elements
    static void getAlternatesRec(int[] arr, int idx, 
                                 ArrayList<Integer> res) {
        if (idx < arr.length) {
            res.add(arr[idx]);
            getAlternatesRec(arr, idx + 2, res);
        }
    }

    static ArrayList<Integer> getAlternates(int[] arr) {
        ArrayList<Integer> res = new ArrayList<>();
        getAlternatesRec(arr, 0, res);
        return res;
    }

    public static void main(String[] args) {
        int[] arr = {10, 20, 30, 40, 50};
        ArrayList<Integer> res = getAlternates(arr);
        for (int x : res) {
            System.out.print(x + " ");
        }
    }
}

File: Data Structure/Array/Array Data Structure Guide/Basic Problems on Array/Alternate elements of an array/Recursive Approach/method1.cpp
// Recursive C++ Program to print alternate elements of the array

#include <iostream>
#include <vector>
using namespace std;

// Recursive function to store all alternate elements
void getAlternatesRec(vector<int> &arr, int idx, vector<int>& res) {
    if(idx < arr.size()) {
        res.push_back(arr[idx]);
        getAlternatesRec(arr, idx + 2, res);
    }
}

vector<int> getAlternates(vector<int> &arr) {
    vector<int> res;
    getAlternatesRec(arr, 0, res);
    return res;
}

int main() {
    vector<int> arr = {10, 20, 30, 40, 50};
    vector<int> res = getAlternates(arr); 
    for(int x: res)
        cout << x << " ";
}